evm_arithmetization/src/cpu/kernel/asm/smt/delete.asm

```diff
@@ -0,0 +1,251 @@
+sibling_is_internal:
+    //stack: sibling_ptr, deleted_child_ptr, bit, node_payload_ptr, level, ks, retdest
+    POP
+insert_child:
+    //stack: deleted_child_ptr, bit, node_payload_ptr, level, ks, retdest
+    %stack (deleted_child_ptr, bit, node_payload_ptr) -> (node_payload_ptr, bit, deleted_child_ptr, node_payload_ptr)
+    ADD %mstore_trie_data
+    // stack: node_payload_ptr, level, ks, retdest
+    %decrement
+    %stack (node_ptr, level, k0, k1, k2, k3, retdest) -> (retdest, node_ptr)
+    JUMP
+
+sibling_is_hash:
+    // stack: sibling_node_type, sibling_ptr, deleted_child_ptr, bit, node_payload_ptr, level, ks, retdest
+    POP
+    //stack: sibling_ptr, deleted_child_ptr, bit, node_payload_ptr, level, ks, retdest
+    %increment %mload_trie_data
+    // stack: hash, deleted_child_ptr, bit, node_payload_ptr, level, ks, retdest
+    %jumpi(insert_child) // Sibling is non-empty hash node.
+sibling_is_empty:
+    // stack: deleted_child_ptr, bit, node_payload_ptr, level, ks, retdest
+    DUP1 %mload_trie_data
+    // stack: deleted_child_node_type, deleted_child_ptr, bit, node_payload_ptr, level, ks, retdest
+    DUP1 %eq_const(@SMT_NODE_HASH) %jumpi(sibling_is_empty_child_is_hash)
+    DUP1 %eq_const(@SMT_NODE_LEAF) %jumpi(sibling_is_empty_child_is_leaf)
+sibling_is_empty_child_is_internal:
+    // stack: deleted_child_node_type, deleted_child_ptr, bit, node_payload_ptr, level, ks, retdest
+    POP
+    // stack: deleted_child_ptr, bit, node_payload_ptr, level, ks, retdest
+    %jump(insert_child)
+
+sibling_is_empty_child_is_hash:
+    // stack: deleted_child_node_type, deleted_child_ptr, bit, node_payload_ptr, level, ks, retdest
+    POP
+    // stack: deleted_child_ptr, bit, node_payload_ptr, level, ks, retdest
+    DUP1 %increment %mload_trie_data
+    // stack: hash, deleted_child_ptr, bit, node_payload_ptr, level, ks, retdest
+    %jumpi(insert_child)
+sibling_is_empty_child_is_empty:
+    // We can just delete this node.
+    // stack: deleted_child_ptr, bit, node_payload_ptr, level, ks, retdest
+    %pop8
+    SWAP1 PUSH 0
+    // stack: retdest, 0
+    JUMP
+
+sibling_is_empty_child_is_leaf:
+    // stack: deleted_child_node_type, deleted_child_ptr, bit, node_payload_ptr, level, ks, retdest
+    POP
+    // stack: deleted_child_ptr, bit, node_payload_ptr, level, k0, k1, k2, k3, retdest
+    %increment
+    // stack: deleted_child_key_ptr, bit, node_payload_ptr, level, k0, k1, k2, k3, retdest
+    DUP4
+    // stack: level, deleted_child_key_ptr, bit, node_payload_ptr, level, k0, k1, k2, k3, retdest
+    DUP3
+    // stack: bit, level, deleted_child_key_ptr, bit, node_payload_ptr, level, k0, k1, k2, k3, retdest
+    PUSH 1 SUB
+    // stack: obit, level, deleted_child_key_ptr, bit, node_payload_ptr, level, k0, k1, k2, k3, retdest
+    DUP3 %mload_trie_data
+    // stack: child_key, obit, level, deleted_child_key_ptr, bit, node_payload_ptr, level, k0, k1, k2, k3, retdest
+    %recombine_key
```

Comment:
Part of another comment, but I don't get why the argument is `obit`, and why you invert the bit again in `recombine_key`.

Reply:
Changed `recombine_key` to not invert the bit.